docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,128 @@
+* Validating they match on concrete instantiation: for example `class Foo<T> where T: struct { }`: `Foo<int>` satisfies constraints, `Foo<object>` doesn't.
+
+Validating on uninstantiated forms is done as part of the old verification logic. The runtime doesn't do this consistently - for example, it is done in calls in IL, but not for ldtoken in IL.
+
+Validating instantiated forms is done always so that a `MethodTable` for an invalid type cannot even be created.
+
+From the runtime perspective, it would be desirable to maintain the invariant that `MethodTable` for a generic type instantiation that doesn't meet its own constraints can never be loaded - an attempt to do so would continue throwing an exception. This has implications on how method bodies that may be violating constraints in certain branches can be compiled. Consider:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator or debugger cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
+
+It would greatly simplify impact of this if we were to limit the amount of places within a method body that are allowed to break the constraints. The least impactful and potentially sufficiently powerful change would be allowing to break constraints on `call` instruction only.
+
+Special attention is needed to the `if` checks that guard constraint-breaking calls. While it would be possible to take the approach in the above examples (have language compilers generate `if` blocks guarded with e.g. `IsAssignableFrom`, or other APIs depending on the form of the constraint), it seems preferable to delegate this responsibility to the runtime itself.
+
+Instead of the language compiler generating `if (typeof(T).IsAssignableFrom(typeof(IInterface))` checks or similar, we'll define an IL sequence that would be intrinsically recognized as a _constraints bridge_. The bridge would take the form of:
+
+```
+call RuntimeHelpers.CheckConstraintsForMethodCallThatFollows
+br.false ....
```

Comment:
Do we really need something like this?

We already know statically if a check might be needed for a given callsite, so why can't we just do the validation implicitly as part of such a `call`?

That is for a specialized generic (value type), we already statically know if a given `T` meets the requirements for a more constrained call. We can simply allow the call or replace it with a throw if it isn't feasible.

Inversely, for a shared generic, we can see that the call might not be possible (we already throw an exception for this today), so we could just replace the throw with an implicit call to `CheckConstraintsForMethodCallThatFollows` and throw if they aren't met.

There isn't really any "need" for it to be explicit in the IR.

Reply:
My general consideration here is that what we're doing is basically no different from bounds checking.

So it is entirely feasible for the JIT/AOT to just insert the implicit constraint check before any call where we need to see if a constraint is met or not.

This is overall less limiting to the IL, will make the IL smaller, and will more easily allow the JIT to hoist, fold, or otherwise optimize these type checks (just like with bounds checking).

----------

This is no different from C# emitting the call itself, the JIT just does it implicitly as it does with bounds checking.

Reply:
We have been trying to maintain ability to verify IL type safety for safe C#.

If we do not introduce a marker for this, we will give up on ability to verify any constraints in IL or have good error handling when the constraints are violated unintentionally (e.g. by mismatch between ref and implementation).

Reply:
C# is already going to have concepts that can't be met via the constraint call (like `unmanaged`) and so much like other new concepts, any verifier may already need to do things like recognize that `if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())` So for IL verification it seems trivial (and goodness) for it recognize these higher level checks already.

While no such need actually exists for IL, which may even include unsafe or unverifiable code (perhaps the user has a private or internal method and already validates the call-sites do the right thing otherwise for example). So it doesn't seem fruitful (to me) to require a lot of additional IR.

----------------------

Consider for example places where a different type check may already exist which C# doesn't quite understand but the user knows is valid. A simple example is in `Vector128<T>` where every call is prefixed with `ThrowHelper.ThrowForUnsupportedIntrinsicsVector128BaseType<T>();` and where the developer knows that this means the `where T : INumber<T>` interface is definitely implemented.

We could make IR even bigger by inserting these runtime check calls, duplicating an extra throw, etc.

But we could also just have the JIT do the proper thing and make this work. Keep IL small, make it easier to optimize the IR (hoisting, folding repeated calls, etc). Make the overall IL more flexible for user patterns (whether verifiable or not), and make it so that the general concept works more generally and not with very stringent constraints around control flow, where the call must exist compared to other calls, ordering, etc

(The current requirements that it must exist before a call with no other calls and only basic operations can even make it very difficult to optimize or hoist other code and may not play well with inlining, helper calls, or other patterns that may be common).

Reply:
> C# is already going to have concepts that can't be met via the constraint call (like unmanaged)

`unmanaged` is outlier that was not implemeted properly accross language/runtime. I agree that we need to consider it in the design, but only as an after-throught to make it work. It should not the element that we design this around.

> Keep IL small

The proposed marker is ~7 bytes of IL in the current proposal. 7 bytes extra for constraint bridging is nothing given overall cost of generic code.

> make it easier to optimize the IR (hoisting, folding repeated calls, etc)

I disagree that the marker prevents optimizations or make optimizations harder.

The marker would be recognized as an intrinsic by the importer (just like the explicit type checks).

For fully specialized generic code, the marked would be imported as a constant that gets folded immediately (just like the explicit type checks).

For shared generic code, the importer would import the marker as constraint-validating dictionary lookup: Fetch the actual call target from the dictionary. If the call target is sentinel value, the constraints are not satisfied and target call is skipped. Otherwise, the constraints are satisfied and the target is called.

In alternative with explicit type checks, the type checks would be separate dictionary lookups and seperate calls. In theory, the JIT can do analysis to prove that these type checks are equivalent to the constraints and fold everything back to a single dictionary lookup. It would be fairly complex optimization: The JIT would have to collect the type conditions checked upfront in some form and pass them to the VM to validate whether the set of conditions is equivalent to the call constraints.


If the marker call that returns bool is undesirable, I would be fine with having a marker call that returns `void` or with having some other explicit identification of constraint bridging calls. However, I think it is important to have explicit non-ambiguous identification of calls that are expected to bridge the generic constraints to avoid guessing whether the type safety violations are intentional or not.

Reply:
> any verifier may already need to do things like recognize that `if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())`

I call out the `unmanaged` constraint specifically - `IsReferenceOrContainsReferences` is not a sufficient check for `unmanaged` constraint because it doesn't do the right thing for `Nullable<T>`. C# would need to generate a call to `IsReferenceOrContainsReferences` _and_ and call to _something_ that is recognized by the verifier as a thing that ensures it meets the `struct, new()` constraint (the only relevant part of `unmanaged` constraint are those two). Whether it's `CheckConstraintsForMethodCallThatFollows` or `default(T) != null`, or something else doesn't matter.

I don't think we need to teach runtimes/verifiers about `unmanaged` constraint for this.

Reply:
> I disagree that the marker prevents optimizations or make optimizations harder.

I'm not sure. Various runtimes (RyuJIT has a few, but Mono has many) hit limitations when trying to make inlining observations due to the introduction of branches, particularly when those branches are in loops or similar.

It's definitely something we can fix longer term, but it's also one of those things where it seems an unnecessary cost to me. That being said, we can at least always decide to support it implicitly in the future if it does turn out to be a problem.

> The proposed marker is ~7 bytes of IL in the current proposal. 7 bytes extra for constraint bridging is nothing given overall cost of generic code.

Notably this will be per call. Users can notably workaround that cost by defining their own helper to containerize the logic, but I expect it's going to be one of those cases where a more typical use-case will result in several calls each with their own 7-byte overhead and the JIT needing to do additional tracking to effectively associate the two calls together.

> For fully specialized generic code, the marked would be imported as a constant that gets folded immediately (just like the explicit type checks).

This isn't really as trivial or even necessarily pay for play when it's explicit as when its implicit. Because its a separate IL opcode and doesn't actually work like an IL prefix, we functionally have to recognize it and then look ahead to at least the next opcode. However, in practice and based on the example, this could potentially be several opcodes ahead (past a branch and past several push/ld opcodes).

We're going to need quite a bit of additional tracking to basically recognize the call, track that we're now expecting a future constrained call to appear "next", fail if any incorrect opcodes appear between the check and the call, then check if the constraint can be met as a constant to determine if the special intrinsic node representing the check should actually be introduced or not. Inversely we could always introduce it, track it, and then remove that node at the call if it's not needed, but in either case this is quite a bit of additional work required in the main import loop to validate the user did everything correctly.

Inversely, making it implicit avoids all of this overhead. The only downside is that you don't know if the user "intended" for it to work that way or not. However, you could "cheat" and make that work with a new IL prefix on the call, using an existing IL prefix with a special marker (say `constrained <RuntimeHelpers.DynamicConstraintCheck> call <methodToken>`), or even some attribute on the method. All of which are slightly more pay for play, less error prone to both produce and consume, and result in smaller IL.

> In alternative with explicit type checks, the type checks would be imported as separate dictionary lookups and separate calls. In theory, the JIT can do analysis to prove that these type checks are equivalent to the constraints and fold everything back to a single dictionary lookup. It would be fairly complex optimization: The JIT would have to collect the type conditions checked upfront in some form and pass them to the VM to validate whether the set of conditions is equivalent to the call constraints.

We're going to have these explicit type checks from C# anyways, some of which are foldable and others (particularly in shared generic code) which won't be. I'm not recommending that the JIT recognize these as making the call safe to do, rather only that it tries to do the call rather than failing statically when the method has opted into that.

It's basically the exact same approach as proposed, just with a less intrusive and less error prone mechanism than inserting a separate `call` before every single more constrained call we want to handle.

Reply:
> Various runtimes (RyuJIT has a few, but Mono has many) hit limitations when trying to make inlining observations due to the introduction of branches, particularly when those branches are in loops or similar.

Yes, we are increasingly depending on great JIT optimizations in the feature designs, investing into RyuJIT heavily and the non-RuyJIT based code generators are falling behind. All the numerics work is a prime example of that. This feature is heading in the same direction.

> I expect it's going to be one of those cases where a more typical use-case will result in several calls each with their own 7-byte overhead and the JIT needing to do additional tracking to effectively associate the two calls together.

The latest version of this proposal is explicitly scoping the type safety violations to calls and prohibits them everywhere else to avoid the issues discussed earlier in this PR. If your code has a string calls, the IL codegen gets complicated as you need to avoid hitting other type safety violations. I expect that the Roslyn codegen will want to always move the whole block to a separate helper method with the more specialized constraint instead of dealing with figuring out whether it is type safe to do everything in inline IL or not.

> We're going to need quite a bit of additional tracking to basically recognize the call, track that we're now expecting a future constrained call to appear "next"

We have a lot of prior art with doing look-ahead like this and it is not particularly complicated. I am not worried.

> `constrained <RuntimeHelpers.DynamicConstraintCheck>`

Nit: This is 6 bytes of IL compared to 5 bytes of IL for void-returning marker call.